### ISSUE TITLE
🐛 Nuisance regression strategies might not be forking properly when despiking is also forking

### ISSUE THREAD

**Issue by sgiavasis** [FCP-INDI/C-PAC#1370]
**Describe the bug**
User reports that not every nuisance regression strategy showed up in the output directory. Note the strategies fork via [Nipype](https://nipype.readthedocs.io/en/0.11.0/users/mapnode_and_iterables.html) [iterables](https://github.com/FCP-INDI/C-PAC/blob/a21841904b66e30f13f37f61dd454ab338aee372/CPAC/pipeline/cpac_pipeline.py#L2114-L2116) (instead of the C-PAC pipeline fork points) and show up as multiple sub-directories under the nuisance-regressed functional outputs.

**To Reproduce**
1. Run a pipeline with multiple nuisance regression strategies (3+ would be good for a test).
2. After the pipeline completes, check the completed nuisance strategies.

**Expected behavior**
There should be a sub-directory for each nuisance regression strategy under the nuisance-regressed functional outputs in the output directory. `functional_nuisance_residuals` is a good directory to check.

**Versions** 
 - C-PAC: 1.7.0
 - Container Platform :
    - Singularity

**Comment by sgiavasis** [comment]
Here is the regressor selection from the user who reported this issue:

> Regressors :
> #1: aCompCor, GSR, no spike regression
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    aCompCor:
>      summary:
>        method: DetrendPC
>        components: 5
>      tissues:
>        - WhiteMatter
>        - CerebrospinalFluid
>      extraction_resolution: 2
>    GlobalSignal:
>      summary: Mean
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
> #2 aCompCor, no GSR, no spike regression
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    aCompCor:
>      summary:
>        method: DetrendPC
>        components: 5
>      tissues:
>        - WhiteMatter
>        - CerebrospinalFluid
>      extraction_resolution: 2
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
> #3 No aCompCor, GSR, no spike regression
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    GlobalSignal:
>      summary: Mean
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
> #4 No aCompCor, no GSR, no spike regression
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
> #5 aCompCor, GSR, spike regression
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    aCompCor:
>      summary:
>        method: DetrendPC
>        components: 5
>      tissues:
>        - WhiteMatter
>        - CerebrospinalFluid
>      extraction_resolution: 2
>    GlobalSignal:
>      summary: Mean
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
>    Censor:
>        method: SpikeRegression
>        thresholds:
>         - type: FD_J
>           value: 0.2
>        number_of_previous_trs_to_censor: 0
>        number_of_subsequent_trs_to_censor: 0
> #6 aCompCor, no GSR, spike regression 
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    aCompCor:
>      summary:
>        method: DetrendPC
>        components: 5
>      tissues:
>        - WhiteMatter
>        - CerebrospinalFluid
>      extraction_resolution: 2
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
>    Censor:
>        method: SpikeRegression
>        thresholds:
>         - type: FD_J
>           value: 0.2
>        number_of_previous_trs_to_censor: 0
>        number_of_subsequent_trs_to_censor: 0
> #7 No aCompCor, GSR, spike regression 
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    GlobalSignal:
>      summary: Mean
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
>    Censor:
>        method: SpikeRegression
>        thresholds:
>         - type: FD_J
>           value: 0.2
>        number_of_previous_trs_to_censor: 0
>        number_of_subsequent_trs_to_censor: 0
> #8 No aCompCor, no GSR, spike regression 
>  - Motion:
>      include_delayed: true
>      include_squared: true
>      include_delayed_squared: true
>    PolyOrt:
>     degree: 2
>    Bandpass:
>      bottom_frequency: 0.01
>      top_frequency: 0.1
>    Censor:
>        method: SpikeRegression
>        thresholds:
>         - type: FD_J
>           value: 0.2
>        number_of_previous_trs_to_censor: 0
>        number_of_subsequent_trs_to_censor: 0

**Comment by shnizzedy** [comment]
I'm able to replicate this issue with the following outputs:

| regressor | label | present in output directory | note |
| --- | --- | --- | --- | 
| `#1: aCompCor, GSR, no spike regression` | `aC-WC-DPC5_G-M_M-SDB_P-2_BP-B0.01-T0.1` | ❌ | aCompCor missing in directory name |
| `#2 aCompCor, no GSR, no spike regression` | `aC-WC-DPC5_M-SDB_P-2_BP-B0.01-T0.1` | ❌ | aCompCor missing in directory name |
| `#3 No aCompCor, GSR, no spike regression` | `G-M_M-SDB_P-2_BP-B0.01-T0.1` | ✔️ ||
| `#4 No aCompCor, no GSR, no spike regression` | `M-SDB_P-2_BP-0.01-T0.1` | ✔️ ||
| `#5 aCompCor, GSR, spike regression` | `aC-WC-DPC5_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2` | ❌ | aCompCor missing in directory name | 
| `#6 aCompCor, no GSR, spike regression` | `aC-WC-DPC5_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2` | ❌ | aCompCor missing in directory name | 
| `#7 No aCompCor, GSR, spike regression` | `G-M_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2` | ✔️ ||
| `#8 No aCompCor, no GSR, spike regression` | `M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2` | ✔️ ||

It seems either aCompCor is not running or is not being labeled (in the latter case, with/without aCompCor would have the same label, so whichever runs second would overwrite the first).

**Comment by shnizzedy** [comment]
The specific user issue here is this forking in concert with forking from

```yaml
runDespike: [0,1]
```

which should generate the 8 outputs above with and without despiking but only one set of 8 outputs is generated

**Comment by sgiavasis** [comment]
Bingo, so then this is really this:
- https://github.com/FCP-INDI/C-PAC/issues/1371
- https://github.com/FCP-INDI/C-PAC/pull/1372

**Comment by sgiavasis** [comment]
Just merged #1372 to develop; can try again.

**Comment by shnizzedy** [comment]
Success!

In the output directory, a subdirectory with despiking and one without:

```tree
.
├── pipeline_debug_nuisance_despike_freq-filter_nuisance
│   └── sub-{subject}_ses-1
|       ├── …
│       ├── functional_nuisance_residuals
│       │   └── _scan_rest_run-1
│       │       ├── _selector_aC-CSF+WM-2mm-DPC5_G-M_M-SDB_P-2_BP-B0.01-T0.1
│       │       ├── _selector_aC-CSF+WM-2mm-DPC5_G-M_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
│       │       ├── _selector_aC-CSF+WM-2mm-DPC5_M-SDB_P-2_BP-B0.01-T0.1
│       │       ├── _selector_aC-CSF+WM-2mm-DPC5_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
│       │       ├── _selector_G-M_M-SDB_P-2_BP-B0.01-T0.1
│       │       ├── _selector_G-M_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
│       │       ├── _selector_M-SDB_P-2_BP-B0.01-T0.1
│       │       └── _selector_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
|       └── …
└── pipeline_debug_nuisance_freq-filter_nuisance
    └── sub-{subject}_ses-1
        ├── …
        ├── functional_nuisance_residuals
        │   └── _scan_rest_run-1
        │       ├── _selector_aC-CSF+WM-2mm-DPC5_G-M_M-SDB_P-2_BP-B0.01-T0.1
        │       ├── _selector_aC-CSF+WM-2mm-DPC5_G-M_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
        │       ├── _selector_aC-CSF+WM-2mm-DPC5_M-SDB_P-2_BP-B0.01-T0.1
        │       ├── _selector_aC-CSF+WM-2mm-DPC5_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
        │       ├── _selector_G-M_M-SDB_P-2_BP-B0.01-T0.1
        │       ├── _selector_G-M_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
        │       ├── _selector_M-SDB_P-2_BP-B0.01-T0.1
        │       └── _selector_M-SDB_P-2_BP-B0.01-T0.1_C-S-0+0-FD-J0.2
        └── …
```
